Add AEMET conditional station updates

Noltari opened the pull request:
## Proposed change
Allow disabling station data gathering, which implies relying only on forecasted values.
AEMET weather stations aren't very reliable and some users are using local sensors together with [weather template](https://www.home-assistant.io/integrations/weather.template/) to override the current temperature, humidity and/or pressure.
But why aren't AEMET stations very reliable? Because average values of last hour are reported hourly (with a delay of 30-40 minutes...) and some stations would stop reporting data occasionally.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Noltari committed:
aemet: merge abstract_aemet_sensor into sensor

Also remove unneded attribution() function.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
aemet: add boolean to enable/disable gathering data from weather stations

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
> Hey there @michaelarnauts, @bdraco, mind taking a look at this pull request as its been labeled with an integration (`tado`) you are listed as a [codeowner](https://github.com/home-assistant/core/blob/dev/CODEOWNERS#L485) for? Thanks!
> (message by CodeOwnersMention)

@michaelarnauts @bdraco sorry for this, but I accidentally changed a tado file...
Please, ignore this.

Noltari committed:
tests: aemet: fix black checks

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented:
Can anyone remove the `integration:tado` label? :)

Noltari commented:
BTW, AbstractAemetSensor was moved into sensor.py as requested by @MartinHjelmare https://github.com/home-assistant/core/pull/48417#discussion_r602794513

MartinHjelmare commented on homeassistant/components/aemet/__init__.py:
```diff
@@ -33,16 +43,26 @@ async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry):
 
     hass.config_entries.async_setup_platforms(config_entry, PLATFORMS)
 
+    update_listener = config_entry.add_update_listener(async_update_options)
```
We can use `hass.config_entries.async_on_unload` to handle the unsubscribe.

MartinHjelmare commented on homeassistant/components/aemet/__init__.py:
```diff
@@ -33,16 +43,26 @@ async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry):
     return True
 
 
+async def async_update_options(hass: HomeAssistant, config_entry: ConfigEntry):
```
```suggestion
async def async_update_options(hass: HomeAssistant, config_entry: ConfigEntry) -> None:
```

MartinHjelmare commented on homeassistant/components/aemet/config_flow.py:
```diff
@@ -47,6 +48,35 @@ async def async_step_user(self, user_input=None):
+class OptionsFlowHandler(config_entries.OptionsFlow):
+    """Handle a option flow for AEMET."""
+
+    def __init__(self, config_entry: config_entries.ConfigEntry):
```
```suggestion
    def __init__(self, config_entry: config_entries.ConfigEntry) -> None:
```

MartinHjelmare commented on homeassistant/components/aemet/weather_update_coordinator.py:
```diff
@@ -148,6 +149,8 @@ async def _get_aemet_weather(self):
         return weather
 
     def _get_weather_station(self):
+        if not self._station_updates:
```
It's better to move the check before calling this method.

Noltari committed:
aemet: introduce changes suggested by @MartinHjelmare

- Use async_on_unload()
- Add a couple of missing "-> None"
- Check if station updates are enabled before calling _get_aemet_weather()

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

MartinHjelmare reviewed (approved):
Looks good!

Noltari commented:
> Looks good!

Thanks!

@MartinHjelmare One question, should the `integration: tado` label be removed before merging this?
I accidentally changed a tado file when I created the PR and removed it immediately, but the bot had already added the label...
